### PR TITLE
add functionality to validate Mutual SSL certificate as a Header

### PR DESCRIPTION
**Pull request by tharindu1st**
add functionality to validate Mutual SSL certificate as a Header

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 1]
```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
...
+                log.debug("Client Certificate Validation disable hence returning client certificate from header");
+                return certificateFromHeader;
+            }
+            try {
+                if (!APIUtil.isCertificateExistInTrustStore(certificateFromMessageContext)) {
+                    log.error("Certificate in messageContext not available in truststore");
+                    return null;
+                } else {
+                    return certificateFromHeader;
+                }
```
[the remark this answers is not in the document]
Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: and I think we can move this to above L265.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 2]
```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
...
+                        .replaceAll(APIMgtGatewayConstants.END_CERTIFICATE_STRING, "");
+
+                byte[] bytes = Base64.decodeBase64(base64EncodedCertificate);
+                try (InputStream inputStream = new ByteArrayInputStream(bytes)) {
+                    X509Certificate x509Certificate = X509Certificate.getInstance(inputStream);
+                    if (APIUtil.isCertificateExistInTrustStore(x509Certificate)) {
+                        certificateFromHeader = x509Certificate;
+                    }
+                } catch (IOException | CertificateException | APIManagementException e) {
+                    log.error("Error while Convert into X509Certificate", e);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]:
```suggestion
                    log.error("Error while converting the certificate to X509Certificate", e);
```

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 3]
```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
...
+                base64EncodedCertificate = URLDecoder.decode(base64EncodedCertificate).
+                        replaceAll(APIMgtGatewayConstants.BEGIN_CERTIFICATE_STRING, "")
+                        .replaceAll(APIMgtGatewayConstants.END_CERTIFICATE_STRING, "");
+
+                byte[] bytes = Base64.decodeBase64(base64EncodedCertificate);
+                try (InputStream inputStream = new ByteArrayInputStream(bytes)) {
+                    X509Certificate x509Certificate = X509Certificate.getInstance(inputStream);
+                    if (APIUtil.isCertificateExistInTrustStore(x509Certificate)) {
+                        certificateFromHeader = x509Certificate;
+                    }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: shall we put a debug log for the else case?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoader.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+
+/**
+ * This class used to load new certificate file into Memory
+ */
+public class CertificateReLoader implements Runnable {
+
+    private static final Log log = LogFactory.getLog(CertificateReLoader.class);
+    private static char[] TRUST_STORE_PASSWORD = System.getProperty("javax.net.ssl.trustStorePassword").toCharArray();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoader.java]: can't `System.getProperty("javax.net.ssl.trustStorePassword")` be null?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoaderUtil.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    private CertificateReLoaderUtil() {
+
+    }
+
+    private static long lastUpdatedTimeStamp;
+
+    public static void startCertificateReLoader() {
+
+        CertificateReLoader certificateReLoader = new CertificateReLoader();
+            executor.scheduleAtFixedRate(certificateReLoader, 60, getCertificateReLoaderInterval(), TimeUnit.SECONDS);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoaderUtil.java]: formatting error

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoaderUtil.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    public static void setLastUpdatedTimeStamp(long lastUpdatedTimeStamp) {
+
+        CertificateReLoaderUtil.lastUpdatedTimeStamp = lastUpdatedTimeStamp;
+    }
+
+    public static long getCertificateReLoaderInterval() {
+
+        String certificateReloaderPeriod =
+                ServiceReferenceHolder.getInstance().getAPIManagerConfigurationService().getAPIManagerConfiguration()
+                        .getFirstProperty(APIConstants.CertificateReLoaderConfiguration.PERIOD);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoaderUtil.java]: config seems incorrect.

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/config/api-manager.xml [hunk 7]
```diff
@@ -760,5 +760,11 @@
     <JWTAuthentication>
         <EnableSubscriptionValidationViaKeyManager>false</EnableSubscriptionValidationViaKeyManager>
     </JWTAuthentication>
-
+    <CertificateReLoaderConfiguration>
+        <Period>10</Period>
+    </CertificateReLoaderConfiguration>
+    <MutualSSL>
+        <ClientCertificateHeader>X-WSO2-CLIENT-CERTIFICATE</ClientCertificateHeader>
+        <DisableClientCertificateValidation>false</DisableClientCertificateValidation>
```
Comment (bhathiya) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/config/api-manager.xml]: shall we change this to enable to match with `apimgt.mutual_ssl.enable_client_validation`?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 8]
```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
...
+            } catch (APIManagementException e) {
+                throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                        "Error while Convert into X509Certificate", e);
+            }
+        } else {
+            return certificateFromMessageContext;
+        }
+    }
+
+    private boolean isClientCertificateValidationDisable() {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: shall we change this to enable to match with `apimgt.mutual_ssl.enable_client_validation`?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 9]
```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
...
+                    log.error("Error while Convert into X509Certificate", e);
+                    throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                            "Error while Convert into X509Certificate", e);
+                }
+            }
+        }
+        // Certificate available in header
+        if (certificateFromHeader != null) {
+            if (isClientCertificateValidationDisable()){
+                log.debug("Client Certificate Validation disable hence returning client certificate from header");
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: Client Certificate Validation disabled. Hence returning client certificate from header"

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java [hunk 10]
```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
...
+            try {
+                if (!APIUtil.isCertificateExistInTrustStore(certificateFromMessageContext)) {
+                    log.error("Certificate in messageContext not available in truststore");
+                    return null;
+                } else {
+                    return certificateFromHeader;
+                }
+            } catch (APIManagementException e) {
+                throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                        "Error while Convert into X509Certificate", e);
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java]: Error while **converting** ...

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoader.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            if (lastUpdatedTimeStamp != lastModified) {
+                CertificateReLoaderUtil.setLastUpdatedTimeStamp(lastModified);
+                localTrustStoreStream = new FileInputStream(trustStoreFile);
+                KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+                trustStore.load(localTrustStoreStream, TRUST_STORE_PASSWORD);
+                ServiceReferenceHolder.getInstance().setTrustStore(trustStore);
+            }
+
+        } catch (KeyStoreException | CertificateException | IOException | NoSuchAlgorithmException e) {
+            log.error("Certificate Couldn't find", e);
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/reloader/CertificateReLoader.java]: Unable to find the certificate

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 12]
```diff
@@ -10087,4 +10097,37 @@ public static boolean isDefaultApplicationCreationDisabledForTenant(int tenantId
...
         return state;
     }
+
+    /**
+     * Validate Certificate exist in TrustStore
+     * @param certificate
+     * @return true if certificate exist in truststore
+     * @throws APIManagementException
+     */
+    public static boolean isCertificateExistInTrustStore(X509Certificate certificate) throws APIManagementException {
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: Exist**s**

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/common/MutualSSLCertificateHandler.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            if (x509Certificates.length > 0) {
+                certificateFromMessageContext = x509Certificates[0];
+            }
+        }
+        if (certificateFromMessageContext != null) {
+            if (!headers.containsKey(Utils.getClientCertificateHeader())) {
+                try {
+                    byte[] base64Encoded = Base64.encodeBase64(certificateFromMessageContext.getEncoded());
+                    String base64EncodedString =
+                            APIMgtGatewayConstants.BEGIN_CERTIFICATE_STRING.concat(new String(base64Encoded)).concat(
```
[the remark this answers is not in the document]
Comment (tharindu1st) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/common/MutualSSLCertificateHandler.java]: this was handle in APIMgtGatewayConstants.BEGIN_CERTIFICATE_STRING

[the remark this answers is not in the document]
Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/common/MutualSSLCertificateHandler.java]: shall we follow one approach for both for consistency?

**Comment by tharindu1st** [comment]
Resolves https://github.com/wso2/product-apim/issues/7418

**Comment by tharindu1st** [comment]
Testcases https://github.com/wso2/product-apim/pull/7419

**Pull request merged**